Chore: refactor docblock parsing into a more specific `DocBlock` value object with an associated registry

Ocramius opened the pull request.

Ocramius committed:
Initial example: isolated some minimal `DocBlock` operations into a value type

This is just demonstrating the feasibility of moving some docblock parsing
out of `PHPUnit\Util\Test`, so we can refactor and cache annotation operations
where applicable.

Ocramius committed:
Extracted `@expectedException` parsing into the `DocBlock` value type

Ocramius committed:
Extracted `@testWith` and `@dataProvider` parsing into the `DocBlock` value type

Ocramius committed:
Removed unused `PHPUnit\Util\Test::parseRequirements()` API

This was moved to the new `DocBlock` API.

Ocramius committed:
Moved inline annotation parsing to the `DocBlock` type

Ocramius committed:
Removed unused constants

Ocramius committed:
Moved parsing of (generic) docblock annotations into `DocBlock`

Ocramius committed:
Moved docblock-level checking of whether a method is a test into `DocBlock`

Ocramius committed:
Extracted pre/post dockblock-controlled hook detection into `DocBlock` API

Ocramius committed:
Introduced an intermediate singleton registry to memoize docblocks

Ocramius committed:
Avoiding (de-)serialization of reflection by extracting everything at construct

Ocramius committed:
Removed unused `isClass` flag

Ocramius committed:
Removed the need to handle `ReflectionFunctionAbstract` in `DocBlock` ctor

Instead, relying on either `ReflectionMethod` or `ReflectionClass` is more
than enough.

Ocramius committed:
`DocBlock#$className` is never nullable according to current semantics

Ocramius committed:
Corrected simple static analysis issues in `DocBlock` implementation

Ocramius committed:
Allow passing `Throwable` to `PHPUnit\Framework\Exception\Exception::__construct()`

Ocramius committed:
Cleaned up and documented `DocBlock#requirements()` return types

A TODO has been introduced, but needs https://github.com/phar-io/version/issues/16
to be fixed first.

Ocramius committed:
Corrected type errors introduced by the stricter `DocBlock` API declaration

Ocramius committed:
Applied automated CS fixes

Ocramius commented on src/Annotation/Registry.php:
```diff
@@ -0,0 +1,90 @@
+ * @internal This class is part of PHPUnit internals, an not intended
+ *           for downstream usage
+ *
+ * @TODO test me (srsly!)
```
Yeap.

Ocramius commented on src/Annotation/DocBlock.php:
```diff
@@ -0,0 +1,561 @@
+                    ];
+                    $recordedOffsets[$matches['name'] . '_constraint'] = $offset;
+                } catch (\PharIo\Version\Exception $e) {
+                    /* @TODO this catch is currently not valid, see https://github.com/phar-io/version/issues/16 */
```
See https://github.com/phar-io/version/issues/16

Ocramius committed:
Corrected type declarations around `PHPUnit\Framework\TestCase#(get|set)Tests` parameters

These were supposed to be `TestCase` rather than more generic `Test` instances.

Ocramius committed:
Documented known type check caused by https://github.com/phar-io/version/issues/16

sebastianbergmann commented:
Looks like we found a bug in @muglug's Psalm:
```
Notice: Undefined property: DOMComment::$tagName in phar:///github/workspace/tools/psalm/src/Psalm/ErrorBaseline.php on line 84

Fatal error: Uncaught Error: Call to undefined method DOMComment::getAttribute() in phar:///github/workspace/tools/psalm/src/Psalm/ErrorBaseline.php:85
Stack trace:
#0 phar:///github/workspace/tools/psalm/src/psalm.php(407): _HumbugBox0ece64d7e152\Psalm\ErrorBaseline::read(Object(_HumbugBox0ece64d7e152\Psalm\Internal\Provider\FileProvider), '.psalm/baseline...')
#1 phar:///github/workspace/tools/psalm/psalm(5): require_once('phar:///github/...')
#2 /github/workspace/tools/psalm(14): require('phar:///github/...')
#3 {main}
  thrown in phar:///github/workspace/tools/psalm/src/Psalm/ErrorBaseline.php on line 85
```
https://github.com/sebastianbergmann/phpunit/pull/3836/checks?check_run_id=215062082

Ocramius commented:
Possibly `baseline.xml` containing a comment:

```
    <!-- see https://github.com/phar-io/version/issues/16 -->
```

Ocramius committed:
Removed `baseline.xml` comment - not supported by psalm

Ref: https://github.com/sebastianbergmann/phpunit/pull/3836#issuecomment-529092829

Ocramius committed:
Verified basic registry API behavior

Ocramius committed:
Applied automated CS fixes

Ocramius committed:
Renamed internal recursive array merge operations, referenced ZF BSD licence

Ocramius committed:
Memoizing `DocBlock#requirements()` so that repeated calls do not lead to overhead

Parsing requirements on docblocks seems to be repeated multiple times,
so we need to prevent this from being performed more than necessary.
The regular expressions around this are quite complex, and therefore there
is a performance impact (~2% total runtime on an IO-free test suite).

Ocramius committed:
Optimised tight cycle in `PHPUnit\Util\Test#parseTestMethodAnnotations()`

This method is used very often (10k+ times in a small test suite), so it
needs to be quite efficient, if possible.

Ocramius committed:
Removed greedy `array_merge()` repeated operation: replaced with single variadic call

Ocramius committed:
Removed greedy `array_merge()` in loop in `PHPUnit\Util\Test#getLinesToBeCoveredOrUsed()`

Replaced by a variadic `array_merge()` call

Ocramius committed:
Revert "Corrected type declarations around `PHPUnit\Framework\TestCase#(get|set)Tests` parameters"

This reverts commit 71432bd56f55d329cb2ee074cbe19a74ce2e1eb7.
